### PR TITLE
Support for Accept and Content-Type headers specified in Feign annota…

### PR DESCRIPTION
**Pull request by Felipe444**
…tions

### CODE REVIEW

File: src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java [hunk 1]
```diff
@@ -25,19 +27,36 @@ public boolean isSupported(Method method) {
...
             .body(RequestBodyExtractor.extract(method.getParameters()))
             .requestParameters(RequestParametersExtractor.extractAll(method))
             .pathParameters(PathParametersExtractor.extractAll(method))
             .build();
     }
 
+    @Override
+    public String[] getResponseMediaHeaders(Method method) {
+        return Arrays.stream(method.getAnnotation(DeleteMapping.class).produces())
+            .map(header -> "Content-Type=" + header)
```
Comment (wlodarcp) [inline src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java:46]: Please move string "Content-Type=" to variable

File: src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/PatchMappingMethodsHandler.java [hunk 2]
```diff
@@ -25,19 +27,36 @@ public boolean isSupported(Method method) {
...
+            .toArray(String[]::new);
+    }
+
+    private String[] getRequestMediaHeaders(Method method) {
+        return ArrayUtils.addAll(
+            Arrays.stream(method.getAnnotation(PatchMapping.class).consumes())
+                .map(header -> "Content-Type=" + header)
+                .toArray(String[]::new),
+            Arrays.stream(method.getAnnotation(PatchMapping.class).produces())
+                .map(header -> "Accept=" + header)
```
Comment (wlodarcp) [inline src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/PatchMappingMethodsHandler.java:56]: For "Accept=" same as for "Content-Type="

File: src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java [hunk 3]
```diff
@@ -209,8 +211,21 @@ private static InteractionResponse createInteractionResponse(
     }
 
     private static Map<String, String> mapHeaders(List<Param> headers) {
+        Map<String, List<String>> mappedHeadersWithDuplicates = mapHeadersWithDuplicates(headers);
+
+        if (mappedHeadersWithDuplicates.values().stream().anyMatch(list -> list.size() > 1)) {
+            log.error("Only single value for given header allowed. Single value will be assigned");
```
Comment (wlodarcp) [inline src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java:217]: What about add more precise info about which values are duplicated?

Comment (wlodarcp) [reply src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java:217]: Also - shouldn't it be better to throw exception in such case?

File: src/test/unit/com/hltech/pact/gen/domain/client/feign/sample/MediaTypeFeignClient.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@FeignClient("SpecProvider")
+public interface MediaTypeFeignClient {
+
+    @PostMapping(path = "/", consumes = {
+        MediaType.APPLICATION_JSON_UTF8_VALUE,
+        MediaType.APPLICATION_PDF_VALUE}, produces = {
```
Comment (wlodarcp) [inline src/test/unit/com/hltech/pact/gen/domain/client/feign/sample/MediaTypeFeignClient.java:16]: move `produces` to new line

File: src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy [hunk 5]
```diff
@@ -1,23 +1,8 @@
...
-import com.hltech.pact.gen.domain.client.feign.sample.OptionalResponseFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.PathFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.RequestBodyFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.RequestHeadersFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.RequestParamFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.RequestTypeFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.ResponseBodyFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.ResponseHeadersFeignClient
-import com.hltech.pact.gen.domain.client.feign.sample.InteractionInfoFeignClient
+import com.hltech.pact.gen.domain.client.feign.sample.*
```
Comment (wlodarcp) [inline src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:5]: Was that intended?

File: src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy [hunk 6]
```diff
@@ -264,6 +249,26 @@ class PactFactoryUT extends Specification {
             }
     }
 
+    def "should get media type from feign client"() {
+        when:
+            final Pact pact = pactFactory.createFromFeignClient(MediaTypeFeignClient, 'SpecConsumer', objectMapper)
```
Comment (wlodarcp) [inline src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:254]: Why we need final here?

File: src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java (hunk 1 above)
Comment (Felipe444) [reply src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java:46]: It id hidden in the private method.

File: src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java (hunk 3 above)
Comment (Felipe444) [reply src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java:217]: It is against JSON rule to duplicate keys. Pacts are expressed as JSONs. I was not sure whether to throw exception here or not as maybe we can treat as "any from set". Maybe I can add entry in README.

File: src/test/unit/com/hltech/pact/gen/domain/client/feign/sample/MediaTypeFeignClient.java (hunk 4 above)
Comment (Felipe444) [reply src/test/unit/com/hltech/pact/gen/domain/client/feign/sample/MediaTypeFeignClient.java:16]: I think readability suffers as indent must be like in the line above.

File: src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy (hunk 5 above)
Comment (Felipe444) [reply src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:5]: No, autoformat did that.

File: src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy (hunk 6 above)
Comment (Felipe444) [reply src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:254]: I just copied from above. It doesn't matter, especially in tests.

Comment (wlodarcp) [reply src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:254]: So it can be removed ;)

Comment (Felipe444) [reply src/test/unit/com/hltech/pact/gen/domain/pact/PactFactoryUT.groovy:254]: yeah

File: src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java (hunk 1 above)
Comment (Felipe444) [reply src/main/java/com/hltech/pact/gen/domain/client/annotation/handlers/DeleteMappingMethodsHandler.java]: Used predefined constant

File: src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java (hunk 3 above)
Comment (matek2305) [inline src/main/java/com/hltech/pact/gen/domain/pact/PactFactory.java:217]: imo adding key to log message could help with potential debugging :)

**Pull request merged**